internal/dyngo/events_test.go

```diff
@@ -0,0 +1,365 @@
+		op2.Finish(testOp2Res{})
+		// Called again
+		require.Equal(t, 2, called)
+
+		op3 := dyngo.StartOperation(testOp3Args{}, dyngo.WithParent(op2))
+		op3.Finish(testOp3Res{})
+		// Not called
+		require.Equal(t, 2, called)
+
+		op2 = dyngo.StartOperation(testOp2Args{}, dyngo.WithParent(op3))
+		op2.Finish(testOp2Res{})
+		// Called again
+		require.Equal(t, 3, called)
+
+		// Finish the operation so that it gets disabled and its listeners removed
+		op1.Finish(testOp1Res{})
+
+		op2 = dyngo.StartOperation(testOp2Args{}, dyngo.WithParent(op1))
+		op2.Finish(testOp2Res{})
+		// No longer called
+		require.Equal(t, 3, called)
+	})
+
+	t.Run("data event", func(t *testing.T) {
+		op1 := dyngo.StartOperation(testOp1Args{})
+
+		var called int
+		op1.OnData((*testOp2Data)(nil), func(op *dyngo.Operation, v interface{}) {
+			_ = v.(testOp2Data)
+			called++
+		})
+
+		op1.EmitData(testOp2Data{})
+		require.Equal(t, 1, called)
+
+		op2 := dyngo.StartOperation(testOp2Args{}, dyngo.WithParent(op1))
+		op2.EmitData(testOp2Data{})
+		op2.Finish(testOp2Res{})
+		require.Equal(t, 2, called)
+
+		op3 := dyngo.StartOperation(testOp3Args{}, dyngo.WithParent(op1))
+		op3.EmitData(testOp2Data{})
+		op3.Finish(testOp3Res{})
+		require.Equal(t, 3, called)
+
+		op2 = dyngo.StartOperation(testOp2Args{}, dyngo.WithParent(op1))
+		op3 = dyngo.StartOperation(testOp3Args{}, dyngo.WithParent(op2))
+		op3.EmitData(testOp2Data{})
+		op3.Finish(testOp3Res{})
+		op2.Finish(testOp2Res{})
+		require.Equal(t, 4, called)
+
+		// Finish the operation so that it gets disabled and its listeners removed
+		op1.Finish(testOp1Res{})
+		op1.EmitData(testOp2Data{})
+		// No new call
+		require.Equal(t, 4, called)
+	})
+
+	// TODO(julio): event dispatch through disabled operations
+}
+
+func BenchmarkEvents(b *testing.B) {
+	type (
+		benchOpArgs struct{ s []byte }
+		benchOpData struct{ s []byte }
+		benchOpRes  struct{ s []byte }
+	)
+
+	dyngo.RegisterOperation((*benchOpArgs)(nil), (*benchOpRes)(nil))
+
+	b.Run("emitting", func(b *testing.B) {
+		// Benchmark the emission of events according to the operation stack length
+		for length := 1; length <= 64; length *= 2 {
+			b.Run(fmt.Sprintf("stack=%d", length), func(b *testing.B) {
+				buf := make([]byte, 1024)
+				rand.Read(buf)
+
+				root := dyngo.StartOperation(benchOpArgs{})
+				defer root.Finish(benchOpRes{})
+
+				op := root
+				for i := 0; i < length-1; i++ {
+					op = dyngo.StartOperation(benchOpArgs{}, dyngo.WithParent(op))
+					defer op.Finish(benchOpRes{})
+				}
+
+				b.Run("start event", func(b *testing.B) {
+					unreg := root.Register(dyngo.OnStartEventListener((*benchOpArgs)(nil), func(*dyngo.Operation, interface{}) {}))
+					defer unreg()
+					b.ReportAllocs()
+					b.ResetTimer()
+					for n := 0; n < b.N; n++ {
+						dyngo.StartOperation(benchOpArgs{buf}, dyngo.WithParent(op))
+					}
+				})
+
+				b.Run("start + finish events", func(b *testing.B) {
```

Comment:
Can you please use one-word test names throughout?  Benchmarks and tests?

Reply:
Will do yes.
Closing it as this is a duplicate comment of https://github.com/DataDog/dd-trace-go/pull/1007#discussion_r723382110